plugins/hls-splice-plugin/src/Ide/Plugin/Splice.hs

```diff
@@ -0,0 +1,334 @@
+        { pluginCommands = commands
+        , pluginCodeActionProvider = Just codeAction
+        }
+
+--
+expandInplaceId, expandCommentedId :: CommandId
+expandInplaceId = "expandTHSpliceInplace"
+expandCommentedId = "expandTHSpliceCommented"
+
+inplaceCmdName :: T.Text
+inplaceCmdName = "expand TemplateHaskell Splice (in-place)"
+
+commentedCmdName :: T.Text
+commentedCmdName = "expand TemplateHaskell Splice (comented-out)"
+
+commands :: [PluginCommand]
+commands =
+    [ PluginCommand expandInplaceId inplaceCmdName $ expandTHSplice Inplace
+    -- , PluginCommand expandCommentedId commentedCmdName $ expandTHSplice Commented
+    ]
+
+expandTHSplice ::
+    -- | Inplace?
+    ExpandStyle ->
+    CommandFunction ExpandSpliceParams
+expandTHSplice eStyle lsp ideState params@ExpandSpliceParams {..} =
+    fmap (fromMaybe defaultResult) $
+        runMaybeT $ do
+            fp <- MaybeT $ pure $ uriToNormalizedFilePath $ toNormalizedUri uri
+            TcModuleResult {tmrParsed = pm, ..} <-
+                MaybeT $
+                    runAction "expandTHSplice.TypeCheck" ideState $
+                        use TypeCheck fp
+            ps <-
+                MaybeT $
+                    useAnnotatedSource
+                        "expandTHSplice.AnnotedSource"
+                        ideState
+                        fp
+            hscEnvEq <-
+                lift $
+                    runAction "expandTHSplice.ghcSessionDeps" ideState $
+                        use_ GhcSessionDeps fp
+            let hscEnv = hscEnvWithImportPaths hscEnvEq
+                srcSpan = rangeToRealSrcSpan range $ fromString $ fromNormalizedFilePath fp
+            ((warns, errs), mEdits) <- calculateEdits lsp ps hscEnv tmrTypechecked srcSpan eStyle params
+            unless (null errs) $
+                reportEditor
+                    lsp
+                    MtError
+                    [ "Error during expanding splice:"
+                    , T.pack $ show errs
+                    ]
+            guard $ null errs
+            unless (null warns) $
+                reportEditor
+                    lsp
+                    MtWarning
+                    [ "Warning during expanding splice:"
+                    , T.pack $ show warns
+                    ]
+            pure
+                ( Right Null
+                , mEdits <&> \edits ->
+                    (WorkspaceApplyEdit, ApplyWorkspaceEditParams edits)
+                )
+    where
+        defaultResult = (Right Null, Nothing)
+
+data SomeHasSplice where
+    MkSomeHasSplice :: HasSplice ast => Proxy# ast -> SomeHasSplice
+
+class (Outputable (ast GhcRn), ASTElement (ast GhcPs)) => HasSplice ast where
+    type SpliceOf ast :: Kinds.Type -> Kinds.Type
+    type SpliceOf ast = HsSplice
+    matchSplice :: Proxy# ast -> ast GhcPs -> Maybe (SpliceOf ast GhcPs)
+    renameSplice :: Proxy# ast -> SpliceOf ast GhcPs -> RnM (Either (ast GhcPs) (ast GhcRn), FreeVars)
```

Comment:
Do we need to rename the splice? Isn't the parsed version good enough for pretty printing?

Reply:
Yes. But, AFAIK, GHC API only provides splice expansion functions with result type in either `GhcRn` or `GhcTc` phases (provided in `RnSplice` and `TcSplice` resp).
Since splicing is taken place in renaming ang typing phase, this design seems reasonable.
Only exception was Pat, as `rnSplicePat` returns `Either GhcPs GhcRn`.

I used `rnSplice*` family provided in `RnSplice` to expand splices, so I named corresponding method as  `renameSplice`, but considering what it actually do, perhaps I should have named it `expandSplice`.

Reply:
OK. The MetaHook should allow you to access the parsed representation of the splice, also avoiding any overhead from having to re-run splices.

Reply:
That sounds great! It seems I should definitely give a try to `MetaHook`. Thanks!